File: clients/android/MPTD/src/com/team4/mptd/Monster.java
package com.team4.mptd;

import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Rect;

public class Monster extends GameObject {

	private double currentHealth;
	private double maxHealth = 100;
	private double directionX;
	private double directionY;
	private double speed = 1;
	private double currDistance = 0;
	private double monsterDmg;
	private boolean isAlive;
	private boolean killedByPlayer = false;
	private GamePath newPath;
	private GamePathPoint lastPoint;
	private Rect healthRect;
	private double hpMod;
	private double armorMod;
	private double speedMod;
	private int cost;
	private int ownerID;

	// param list: the int type of the monster, the sprite of the monster, the
	// path of the monster, the GameSession instance, frame width, frame height,
	// number of frames
	public Monster(double hpMod, double armorMod, double speedMod,
			Bitmap sprite, GamePath path, int frWidth, int frHeight,
			int noFrames) {
		super(path.pathStart.p.x - frWidth / 2, path.pathStart.p.y - frHeight
				/ 2, frWidth, frHeight, noFrames);

		this.hpMod = hpMod;
		this.armorMod = armorMod;
		this.speedMod = speedMod;
		
		// COST OF MONSTER HARD CODED
		cost = 2;

		maxHealth = maxHealth * hpMod;
		currentHealth = maxHealth;
		speed = speed * speedMod;
		isAlive = true;
		newPath = path;
		lastPoint = newPath.pathStart;
		currDistance = lastPoint.length;
		directionX = lastPoint.dir.x;
		directionY = lastPoint.dir.y;
		if (sprite != null) {
			setBitmap(sprite);
		}
		monsterDmg = 1;

	}
	
	private void updateStateID() {
		if(directionX == 0) {
			if(directionY == 1) { //direction is down
				stateID = 1;
			} else { // direction is up
				stateID = 0;
			}
		} else {
			if(directionX == 1) { // direction is right
				stateID = 2;
			} else { // direction is left
				stateID = 3;
			}
		}
	}

	@Override
	public void draw(Canvas canvas, Rect src, float offsetX, float offsetY) {
		super.draw(canvas, src, offsetX, offsetY);

		// Draw health bar
		healthRect = new Rect(thisRect.left - (int) offsetX, thisRect.top - 20
				- (int) offsetY, thisRect.left
				+ (int) ((currentHealth / maxHealth) * frameWidth)
				- (int) offsetX, thisRect.top - 10 - (int) offsetY);
		Paint p = new Paint();
		p.setARGB(255, 127, 255, 0);
		canvas.drawRect(healthRect, p);
	}
	
	public int getCost() {
		return cost;
	}

	public double getDmg() {
		return monsterDmg;
	}

	public boolean isAlive() {
		return isAlive;
	}

	public void setSpeed(double sp) { // sets the speed of the monster
		speed = sp;
	}

	public double getHealth() { // returns the current health of the monster
		return currentHealth;
	}

	public double getMaxHealth() {
		return maxHealth;
	}

	public double getSpeed() {
		return speed;
	}
	public void setOwner(int owner){
		ownerID = owner;
	}
	public int getOwner(){
		return ownerID;
	}

	public void takeDamage(double dmg) { // applies damage taken to monster
											// health
		currentHealth = currentHealth - dmg;
		if (currentHealth <= 0) {
			session.monsterWasKilled(this);
			session.getPlayer().defeatedMonster();
			isAlive = false;
			killedByPlayer = true;
		}
	}

	public boolean isKilledByPlayer() {
		return killedByPlayer;
	}

	// Updates the status of the monster
	public void update() {
		// update the position of the monster
		move();
	}

	// changes the monsters position
	public void move() {
		// Move monster position distance governed by its speed.
		moveDistance(speed);
		
		// If monster just passed the next waypoint
		if (currDistance <= 0) {

			lastPoint = lastPoint.next;
			if (lastPoint.next == null) {
				// monster passed last point, time to despawn.
				isAlive = false;
				session.monsterPassedThrough(this);
				return;
			}
			// set new left, top to first waypoint coordinates and direction.
			left = lastPoint.p.x - frameWidth / 2; // compensate for left and
													// top reference benig
													// corner of monster frame
			top = lastPoint.p.y - frameHeight / 2;
			directionX = lastPoint.dir.x;
			directionY = lastPoint.dir.y;

			// time to move the remaining distance in the new direction
			// we always do a remainder move since we rarely will have the
			// condition currDistance == 0.
			moveDistance(-currDistance);
			// Set currDistance to next length
			currDistance = lastPoint.length + currDistance;
		}

		// update the position of the monster
		updateRect();
		// update state direction of the monster, used by animation to select correct row.
		updateStateID();
	}

	// unconditionally adjusts the monsters position
	public void moveDistance(double distance) {
		double distX = distance * directionX;
		double distY = distance * directionY;
		left = left + distX;
		top = top + distY;
		// Adjusts the currDistance counter.
		currDistance = currDistance - distance;
	}
	
	public void setCost(int cost) {
		this.cost = cost;
	}

	@Override
	public Monster clone() {
		Monster mon = new Monster(hpMod, armorMod, speedMod, bmap, newPath,
				frameWidth, frameHeight, noOfFrames);
		mon.setTypeID(this.getTypeID());
		
		return mon;
	}
}

File: clients/windesktop/C-SharpClient_1.1/C-SharpClient_1.1/GameObjects/Monster.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Audio;
using Microsoft.Xna.Framework.Content;
using Microsoft.Xna.Framework.GamerServices;
using Microsoft.Xna.Framework.Graphics;
using Microsoft.Xna.Framework.Input;
using Microsoft.Xna.Framework.Media;

namespace C_SharpClient_1._1
{
    class Monster : GameObject
    {
        GameSession.GamePath.GamePathPoint lastPoint;
        double currDistance = 0;

        public SpriteFont setHpBar { set { base.font1text = value; } }

        private bool isKilledByPlayer = false;
        public bool IsKileldByPlayer { get { return isKilledByPlayer; } }

        public int damage;
        public double hp;
        public int maxHP;
        private int speed;
        public int maxSpeed;
        public int cost = 0;
        public int ownerID;
        Vector2 position;
        public int Speed { get { return speed; } set { speed = value; } }
        private bool isAlive = true;
        public bool Alive { get { return isAlive; } set { isAlive = value; } }

        /// <summary>
        /// This is just for testing
        /// </summary>
        /// <param name="rec">Rectangle</param>
        /// <param name="hp">Helth Points</param>
        public Monster(Rectangle rec, double hp, int speed)
        {
            base.Rec = rec;
            base.type = "Monster";
            this.speed = speed;
            this.hp = hp;
        }

        ///  To construct a sprite you need this:
        /// </summary>
        /// <param name="txt2D">Textture 2D (Must be Loaded from content)</param>
        /// <param name="X">X position on screen</param>
        /// <param name="Y">Y Position on screen</param>
        /// <param name="map"> Monsters needs a path to walk on the map</param>
        public Monster(Texture2D txt2D, double hp, int speed)
        {
            this.maxSpeed = speed;
            this.maxHP = (int)hp;
            base.type = "Monster";
            this.speed = speed;
            base.Text2D = txt2D;
            base.Rec = new Rectangle(0, 0, txt2D.Width, txt2D.Height);
            this.hp = hp;
        }

        /// <summary>
        /// To construct a sprite you need this:
        /// </summary>
        /// <param name="txt2D">Textture 2D (Must be Loaded from content)</param>
        /// <param name="rec">Rectangle have both wigth heigth and Position</param>
        public Monster(Texture2D txt2D, Rectangle rec, double hp, int speed)
        {
            base.Text2D = txt2D;
            base.Rec = rec;
            this.maxSpeed = speed;
            this.maxHP = (int)hp;
            base.type = "Monster";
            this.speed = speed;
            this.hp = hp;
        }


        /// <summary>
        /// Lower the monster HP
        /// </summary>
        /// <param name="damage">The ammoint of damage the monster takes</param>
        /// <returns>Returns HP left</returns>
        public double ReduceHP(double damage)
        {
            hp -= damage;
            if (hp <= 0)
            {
                isAlive = false;
                isKilledByPlayer = true;
            }
            return hp;
        }

        private int nextFrame = 0;
        /// <summary>
        /// The Movement Speed of the monster
        /// </summary>
        /// <returns>The current speed of the monster</returns>
        public override bool Update(ref List<GameObject> listToPrint)
        {
            nextFrame++;
            if (nextFrame > 1)
            {
                nextFrame = 0;
                if (Math.Round(this.lastPoint.dir.X) == -1)
                    SetAnimation(3);
                if (Math.Round(this.lastPoint.dir.X) == 1)
                    SetAnimation(2);
                if (Math.Round(this.lastPoint.dir.Y) == 1)
                    SetAnimation(1);
                if (Math.Round(this.lastPoint.dir.Y) == -1)
                    SetAnimation(0);
            }

            base.strouput = Convert.ToString(hp);
            if (this.lastPoint != null)
                move();
            return !isAlive;
        }

        public Monster Clone()
        {
            return (Monster)this.MemberwiseClone();
        }
        public override void SetGameSession(GameSession gameSession)
        {
            base.gameSession = gameSession;
            this.lastPoint = gameSession.gamePath.PathStart;
            currDistance = this.lastPoint.length;
            position = this.lastPoint.p;
            position.X -= 25F;
            position.Y -= 25F;
            base.Rec = new Rectangle((int)lastPoint.p.X -25, (int)lastPoint.p.Y -25, base.Rec.Width, base.Rec.Height);
        }

        private int currentFrame = 0;
        private void SetAnimation(int stateID)
        {
            currentFrame++;
            if (currentFrame > (base.Text2D.Width / base.ScaleRectangle.Width) - 1)
                currentFrame = 0;
            // just update to next frame
            base.ScaleRectangle = new Rectangle((currentFrame * base.ScaleRectangle.Width), (stateID * base.ScaleRectangle.Height), base.ScaleRectangle.Width, base.ScaleRectangle.Height);
        }

        // changes the monsters position
        public void move()
        {
            // Move monster position distance governed by its speed.
            moveDistance(speed);

            // If monster just passed the next waypoint
            if (currDistance <= 0)
            {
                if (lastPoint.next == null)
                {
                    // monster passed last point, time to despawn.
                    isAlive = false;
                    return;
                }
                lastPoint = lastPoint.next;
                // set new left, top to first waypoint coordinates and direction.
                position = this.lastPoint.p;
                position.X -= 25F;
                position.Y -= 25F;
                base.Rec = new Rectangle((int)this.lastPoint.p.X -25, (int)this.lastPoint.p.Y -25, base.Rec.Width, base.Rec.Height);

                moveDistance(-1 * currDistance);
                // Set currDistance to next length
                currDistance = lastPoint.length + currDistance;
            }
        }

        // unconditionally adjusts the monsters position
        public void moveDistance(double distance)
        {
            position.X += (float)distance * this.lastPoint.dir.X;
            position.Y += (float)distance * this.lastPoint.dir.Y;
            base.Rec = new Rectangle((int)position.X , (int)position.Y , base.Rec.Width, base.Rec.Height);
            // Adjusts the currDistance counter.
            currDistance = currDistance - distance;
        }
    }
}
